Further separate out OS abstraction / GLib and add an embedded abstraction

mmlr opened the pull request:
In the process of integrating FluidSynth into BespokeSynth, I have been bitten by the GLib dependency. While everything happened to work out of the box for me on Linux, adding the GLib dependency for the Windows and macOS builds seemed daunting and unnecessary. So I, as many before me, set out to strip the GLib dependency. I have studied the discussion in #847 carefully to possibly come up with something that will not be immediately dismissed.

I've essentially picked up on [the proposal in this comment](https://github.com/FluidSynth/fluidsynth/discussions/847#discussioncomment-807707) to further separate out GLib specifics and make adding a different abstraction easier.  Due to a lot of prior work in abstraction to fluid_sys.*, especially in #901, the actual work was reasonably simple to do. There were some non-GLib dependent things in fluid_sys that remain, while anything GLib dependent is moved to fluid_sys_glib.

With this new setup in place, I've implemented an abstraction specifically meant for pure embedding uses. It is mostly stubbed out and does not support threading at all, doesn't provide proper atomics or time functions, etc. While it still can be used with the standalone fluidsynth application for file rendering, all the advanced features will simply error out. But it allows essentially dependency free inclusion into other projects where only the subset of audio rendering is needed. From what I gather in the discussion, this is a common use case that previously made FluidSynth in its current form unattractive for other projects and therefore IMHO warrants this abstraction.

Special care has been taken to make it obvious when stubbed functionality is being used (by means of a logged error message) so that it should be obvious what is going on when the stubs are hit and things fail. Further documentation would certainly help to facilitate possible other abstractions, although the interface is already pretty straight forward and one can easily base a new abstraction on the embedded one. If this approach is deemed worthwhile, I'd be willing to put that effort into full documentation and possibly other abstractions.

I have tried to separate out the commits so that they stand on their own. The cleanup could also be applied separately, but the rest build on top of each other.

derselbst reviewed (commented):
I like that - It's a step in the right direction, thanks!

Got a few nit-picks and questions below.

mmlr commented on CMakeLists.txt:
```diff
@@ -103,6 +103,10 @@ option ( enable-threads "enable multi-threading support (such as parallel voice
 
+set ( os-abstraction "glib" CACHE STRING "OS abstraction to use, provided by src/utils/fluid_sys_${os-abstraction}.*" )
+
+add_definitions ( -D OS_ABSTRACTION_${os-abstraction} )
```
I'll take a look. Right now it explicitly checks for one of the known values and therefore requires the macro to be set. It does that so that it can error out explicitly when not configured instead of assuming a default. But that can easily be changed so it just defaults to GLib if unset.

mmlr replied:
Added to config.cmake. Could have used `string ( TOUPPER ${osal} OSAL )` to make the macro read OSAL_GLIB instead of OSAL_glib. Shall I change to that or would that again pollute the global namespace if done in the top 
 level CMakeLists?

mmlr commented on src/utils/fluidsynth_priv.h:
```diff
@@ -31,8 +31,18 @@
 
+#undef FLUID_LIKELY
+#undef FLUID_UNLIKELY
+#define FLUID_LIKELY G_LIKELY
+#define FLUID_UNLIKELY G_UNLIKELY
```
They can, it just means that all the places they are used need to include fluid_sys.h instead of just fluidsynth_priv.h. I wasn't sure which way around was preferred.

derselbst replied:
Yeah, right... Out of scope for this PR, let's keep it for now.

derselbst commented on src/utils/fluid_sys.h:
```diff
@@ -112,23 +112,18 @@
+#if OS_ABSTRACTION_glib
+#include "fluid_sys_glib.h"
+#elif OS_ABSTRACTION_embedded
+#include "fluid_sys_embedded.h"
```
Should be possible with https://cmake.org/cmake/help/latest/module/CMakeDependentOption.html#module:CMakeDependentOption

mmlr replied:
Just made the default setting change for now.

derselbst commented on src/rvoice/fluid_iir_filter.cpp:
```diff
@@ -20,6 +20,7 @@
 
 #include "fluid_iir_filter.h"
 #include "fluid_conv.h"
+#include "fluid_sys.h"
```
I don't quite understand this part. Perhaps if you could share an error message when not including fluid_sys here would be helpful to understand. I'm currently unable to compile myself, sry.

mmlr replied:
The problem is the TRUE/FALSE assigned to calc_coeff_flag in [lines 295](https://github.com/FluidSynth/fluidsynth/blob/30611650e24475823d2ee4544c6f9cd9196157bc/src/rvoice/fluid_iir_filter.cpp#L295), 334 and 356. The [TRUE/FALSE](https://docs.gtk.org/glib/macros.html#values) are defined via the `glib.h` include in `fluidsynth_priv.h` when building with GLib, resulting in `error: 'FALSE' was not declared in this scope` when not building with GLib.

As they are GLib specific, they are now part of the OSAL and for `embedded` they come from `fluid_sys_embedded.h` and aren't available when only including `fluidsynth_priv.h`.

As the calc_coeff_flag is actually no gboolean at all but an unsigned int, using 1/0 instead or making it an actual bool and using true/false would also solve this. The latter seems the cleanest, so I'll simply do that and drop the `fluid_sys.h` include again.

derselbst replied:
Oh, now I see. Ofc, using `bool` is totally valid in C++.

mmlr commented:
> LGTM now, thanks! I'll keep it around here for a few days, in case anybody else wants to provide feedback.

Sounds great, thanks. If I make some further improvements on this, should I add them here or make new PRs later?

> (Could you pls. `git merge origin/master` to resolve the renaming conflict?)

I've rebased on top of master and squashed some of the intermediate commits to make the history more readable. The final state is the same as before.

derselbst commented:
> Sounds great, thanks. If I make some further improvements on this, should I add them here or make new PRs later?

I'd prefer a separate PR. Could be a stacked PR, in case you're faster than me merging this one.

mmlr commented:
I have rebased the change on top of current master and fixed a wrong sizeof() for an allocation by using FLUID_NEW. I've added it in this PR because IMO there's no value in having the wrong version merged. In the meantime I've also implemented a POSIX OSAL version that I will add as a separate PR.

derselbst commented:
> I have rebased the change on top of current master and fixed a wrong sizeof() for an allocation by using FLUID_NEW. I've added it in this PR because IMO there's no value in having the wrong version merged.

That's fine.

> In the meantime I've also implemented a POSIX OSAL version that I will add as a separate PR.

I am a bit reluctant to accept platform-specific OSAL's. It would open the door for accepting numerous other OSALs, starting with Windows, and it would ultimately get fluidsynth back into a pre-glib state, where multiple (possibly buggy) OSAL had to be maintained in upstream code.

IMO, a better way to go forward would be a C++11 based OSAL, as it has wider support than e.g. C11.

mmlr commented:
> I am a bit reluctant to accept platform-specific OSAL's. It would open the door for accepting numerous other OSALs, starting with Windows, and it would ultimately get fluidsynth back into a pre-glib state, where multiple (possibly buggy) OSAL had to be maintained in upstream code.

I understood as much from the modernization discussion. I agree in principle and for my specific use-case, I don't really need anything beyond a dependency free embedded OSAL as implemented here.

Generally having the option of not pulling in a big dependency seems worthwhile. But as per the discussion, to get a full FluidSynth, including libinstpatch and LADSPA support, you currently need GLib anyway. So these abstractions can currently only provide so much, outside of use-cases where you maybe don't need or want these features.

> IMO, a better way to go forward would be a C++11 based OSAL, as it has wider support than e.g. C11.

I've [implemented an initial version of this](https://github.com/mmlr/fluidsynth/commit/400776a5ebb06686505cdd4bb4c5985b64c590e9). From a threading and locking perspective, this should fully work. The atomics are suboptimal but should have correct semantics.

Really the sticking points are file access and shell parsing. The shell parsing needs few features and could easily be internalized. File checks and modification time could use C++17 filesystem, but I have no idea how widely this is supported.

mmlr commented:
> > IMO, a better way to go forward would be a C++11 based OSAL, as it has wider support than e.g. C11.
> 
> Really the sticking points are file access and shell parsing. The shell parsing needs few features and could easily be internalized. File checks and modification time could use C++17 filesystem, but I have no idea how widely this is supported.

I went ahead and did both of these, see [my osal-cpp11 branch](https://github.com/mmlr/fluidsynth/compare/os-abstraction...osal-cpp11). For the missing file system abstraction, C++17 is automatically used when it is available. It turns out that [this seems to be broadly supported](https://en.cppreference.com/w/cpp/compiler_support/17). A minimal compliant version (~200 LoC) of shell parsing is also provided internally (with a unit test to validate it against what the GLib version outputs). Except for LADSPA and libinstpatch support, this makes the C++ OSAL functionally complete.

Overall the OSAL provided in this PR seems to cover all that is needed to write useful abstractions.

derselbst commented:
Ok, wow. I'll go ahead and merge this now. The question of whether to use C++11 vs. C++17 should be once again raised in #847. I am aware of users who target Win98 and WinXP. While I am willing to drop Win98 compatibility, I preferred to go with C++11 as it can be compiled for WinXP. The technical advantages of C++17 are obvious though, no only for the std library, but also compiler-wise for porting the lookup tables to `constexpr`.

Yet, feel free to open a PR already for that branch. I'd be curious to see which CI pipeline(s) fails to compile with the cpp11 osal activated.

Thanks again!
